README.md
# svg-to-image-proxy

[![version](https://img.shields.io/npm/v/svg-to-image-proxy.svg?style=flat-square)][npm]
[![license](https://img.shields.io/npm/l/svg-to-image-proxy.svg?style=flat-square)][npm]
[![build](https://img.shields.io/circleci/project/github/badges/svg-to-image-proxy.svg?style=flat-square)][build]
[![code style](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)][prettier]

[npm]: https://npmjs.com/svg-to-image-proxy
[build]: https://circleci.com/gh/badges/svg-to-image-proxy/tree/master
[prettier]: https://prettier.io/

Web function to generate PNGs from SVGs. Based on [convert-svg-to-png][]
which uses headless Chromium.

The proxy accepts a request for e.g. `https://raster.example.com/example.png`
and translates it to a request for
`https://other-server.example.com/example.svg`. If the path doesn't contain
`.png`, it passes is unchanged, e.g. `https://raster.example.com/example` to
`https://other-server.example.com/example`.

The query string is also forwarded upstream. Cache headers are forwarded downstream.

The host comes from the `BASE_URL` environment variable.

Designed for [Shields][] though may be useful for other application as well.

[shields]: https://github.com/badges/shields
[convert-svg-to-png]: https://www.npmjs.com/package/convert-svg-to-png

## Deployment

Runs on Zeit Now.

Staging is deployed automatically from master: https://shields-raster-staging.now.sh/

Production is deployed manually for now.

To deploy, run `now alias shields-raster-staging.now.sh shields-raster.now.sh`

The production server is at: https://raster.shields.io/

## License

This project is licensed under the MIT license.


CHANGELOG.md
# Changelog

## 0.2.0

Adapt for Zeit Now.

## 0.1.0

Initial release.


test.js
'use strict'

const { expect } = require('chai')
const got = require('got')
const isPng = require('is-png')
const micro = require('micro')
const nock = require('nock')
const listen = require('test-listen')
const setup = require('.')

const { cleanup } = setup

nock.disableNetConnect()
nock.enableNetConnect(/localhost|127.0.0.1/)

const baseUrl = process.env.BASE_URL

const badgeSvg =
  '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="66" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="a"><rect width="66" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#a)"><path fill="#555" d="M0 0h37v20H0z"/><path fill="#007ec6" d="M37 0h29v20H37z"/><path fill="url(#b)" d="M0 0h66v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110"> <text x="195" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="270">hello</text><text x="195" y="140" transform="scale(.1)" textLength="270">hello</text><text x="505" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="190">bar</text><text x="505" y="140" transform="scale(.1)" textLength="190">bar</text></g> </svg>'

after(async function() {
  await cleanup()
})

describe('svg-to-image-proxy endpoint', function() {
  let service, url
  beforeEach(async function() {
    this.timeout('10s')
    service = micro(await setup)
    url = await listen(service)
  })
  afterEach(function() {
    if (service) {
      service.close()
      service = undefined
    }
  })

  it('rasterizes svg fetched from the expected address', async function() {
    this.timeout('20s')

    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg')
      .reply(200, badgeSvg)

    const { body, statusCode } = await got(`${url}/badge/foo-bar-blue.svg`, {
      encoding: null,
      retry: { retries: 0 },
    })

    expect(body).to.satisfy(isPng)
    expect(statusCode).to.equal(200)

    scope.done()
  })

  it('sets the appropriate content-type header', async function() {
    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg')
      .reply(200, badgeSvg)

    const { headers } = await got(`${url}/badge/foo-bar-blue.svg`, {
      encoding: null,
      retry: { retries: 0 },
    })

    expect(headers['content-type']).to.equal('image/png')

    scope.done()
  })

  it('forwards query params', async function() {
    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg?label=hello')
      .reply(200, badgeSvg)

    await got(`${url}/badge/foo-bar-blue.svg?label=hello`, {
      encoding: null,
      retry: { retries: 0 },
    })

    scope.done()
  })

  it('forwards If-None-Match request header upstream', async function() {
    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg')
      .matchHeader('if-none-match', 'abcde')
      .reply(200, badgeSvg)

    await got(`${url}/badge/foo-bar-blue.svg`, {
      encoding: null,
      retry: { retries: 0 },
      headers: { 'If-None-Match': 'abcde' },
    })

    scope.done()
  })

  it('forwards Cache-Control response header downstream', async function() {
    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg')
      .reply(200, badgeSvg, { 'Cache-Control': 'max-age=0, must-revalidate' })

    const { headers } = await got(`${url}/badge/foo-bar-blue.svg`, {
      encoding: null,
    })

    expect(headers['cache-control']).to.equal('max-age=0, must-revalidate')

    scope.done()
  })

  it('displays an error badge when the upstream server provides invalid SVG', async function() {
    this.timeout(5000)
    const scope = nock(baseUrl)
      .get('/badge/foo-bar-blue.svg')
      // The server tries three times.
      .times(3)
      .reply(500)

    const { body, statusCode } = await got(`${url}/badge/foo-bar-blue.svg`, {
      encoding: null,
      retry: { retries: 0 },
      throwHttpErrors: false,
    })

    expect(statusCode).to.equal(502)
    expect(body).to.satisfy(isPng)

    scope.done()
  })

  it('returns 404 for ignored paths', async function() {
    const { body, statusCode } = await got(`${url}/favicon.ico`, {
      encoding: null,
      retry: { retries: 0 },
      throwHttpErrors: false,
    })

    expect(statusCode).to.equal(404)
    expect(Buffer.alloc(0).equals(body))
  })
})


index.js
'use strict'

const { URL } = require('url')
const { createConverter } = require('convert-svg-to-png')
const got = require('got')
const isSvg = require('is-svg')

const converter = createConverter()
async function toPng(svg) {
  return converter.convert(svg)
}
async function cleanup() {
  await converter.destroy()
}

const baseUrl = process.env.BASE_URL
if (!baseUrl) {
  throw Error('BASE_URL must be set to the server that is hosting the SVGs')
}

const requestHeadersToForward = [
  'if-modified-since',
  'if-unmodified-since',
  'if-none-match',
]

const responseHeadersToForward = [
  'date',
  'cache-control',
  'expires',
  'last-modified',
]

const ignoredPaths = ['/favicon.ico']

async function handler(req, res) {
  res.setHeader('content-type', 'image/png')

  const requestHeaders = {}
  requestHeadersToForward.forEach(header => {
    requestHeaders[header] = req.headers[header]
  })

  // Ensure we're always using the `baseUrl` by using just the path from
  // the request URL.
  const { pathname, search } = new URL(req.url, 'https://bogus.test')

  if (ignoredPaths.includes(pathname)) {
    res.statusCode = 404
    return ''
  }

  const outPath = pathname.replace('.png', '.svg')
  const outUrl = new URL(outPath, baseUrl)
  outUrl.search = search

  const { body: svg, headers: responseHeaders } = await got(outUrl, {
    headers: requestHeaders,
    throwHttpErrors: false,
  })

  responseHeadersToForward.forEach(header => {
    if (header in responseHeaders) {
      res.setHeader(header, responseHeaders[header])
    }
  })

  if (isSvg(svg)) {
    return toPng(svg)
  } else {
    console.log(svg)
    res.statusCode = 502
    return errorBadge
  }
}

let errorBadge
async function setup() {
  errorBadge = await toPng(
    // Fetched from https://img.shields.io/badge/invalid-svg-red.svg
    '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="160" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="a"><rect width="160" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#a)"><path fill="#555" d="M0 0h37v20H0z"/><path fill="#e05d44" d="M37 0h123v20H37z"/><path fill="url(#b)" d="M0 0h160v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110"> <text x="195" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="270">error</text><text x="195" y="140" transform="scale(.1)" textLength="270">error</text><text x="975" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="1130">invalid svg response</text><text x="975" y="140" transform="scale(.1)" textLength="1130">invalid svg response</text></g> </svg>'
  )
  return handler
}
module.exports = setup()
module.exports.cleanup = cleanup


rasterize.js
'use strict'

const { URL } = require('url')
const chrome = require('chrome-aws-lambda')
const { createConverter } = require('convert-svg-to-png')
const got = require('got')
const isSvg = require('is-svg')

const baseUrl = process.env.BASE_URL
if (!baseUrl) {
  throw Error('BASE_URL must be set to the server that is hosting the SVGs')
}

const requestHeadersToForward = [
  'if-modified-since',
  'if-unmodified-since',
  'if-none-match',
]

const responseHeadersToForward = [
  'date',
  'cache-control',
  'expires',
  'last-modified',
]

const ignoredPaths = ['/', '/favicon.ico']

let cleanup
async function setup() {
  const converter = createConverter({
    puppeteer: {
      args: chrome.args,
      executablePath: await chrome.executablePath,
      headless: chrome.headless,
    },
  })
  cleanup = async () => converter.destroy()

  const errorBadge = await converter.convert(
    // Fetched from https://img.shields.io/badge/invalid-svg-red.svg
    '<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="160" height="20"><linearGradient id="b" x2="0" y2="100%"><stop offset="0" stop-color="#bbb" stop-opacity=".1"/><stop offset="1" stop-opacity=".1"/></linearGradient><clipPath id="a"><rect width="160" height="20" rx="3" fill="#fff"/></clipPath><g clip-path="url(#a)"><path fill="#555" d="M0 0h37v20H0z"/><path fill="#e05d44" d="M37 0h123v20H37z"/><path fill="url(#b)" d="M0 0h160v20H0z"/></g><g fill="#fff" text-anchor="middle" font-family="DejaVu Sans,Verdana,Geneva,sans-serif" font-size="110"> <text x="195" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="270">error</text><text x="195" y="140" transform="scale(.1)" textLength="270">error</text><text x="975" y="150" fill="#010101" fill-opacity=".3" transform="scale(.1)" textLength="1130">invalid svg response</text><text x="975" y="140" transform="scale(.1)" textLength="1130">invalid svg response</text></g> </svg>'
  )

  async function handler(req, res) {
    res.setHeader('content-type', 'image/png')

    const requestHeaders = {}
    requestHeadersToForward.forEach(header => {
      requestHeaders[header] = req.headers[header]
    })

    // Ensure we're always using the `baseUrl` by using just the path from
    // the request URL.
    const { pathname, search } = new URL(req.url, 'https://bogus.test')

    if (ignoredPaths.includes(pathname)) {
      res.statusCode = 404
      return ''
    }

    const outPath = pathname.replace('.png', '.svg')
    const outUrl = new URL(outPath, baseUrl)
    outUrl.search = search

    const { body: svg, headers: responseHeaders, statusCode } = await got(
      outUrl,
      {
        headers: requestHeaders,
        throwHttpErrors: false,
      }
    )
    console.log(`Status code ${statusCode}`)

    responseHeadersToForward.forEach(header => {
      if (header in responseHeaders) {
        res.setHeader(header, responseHeaders[header])
      }
    })

    if (isSvg(svg)) {
      return converter.convert(svg)
    } else {
      console.log(`Expected svg, got: ${svg}`)
      res.statusCode = 502
      return errorBadge
    }
  }

  return handler
}
const handlerPromise = setup()

module.exports = async (req, res) =>
  require('micro').run(req, res, async (req, res) =>
    (await handlerPromise)(req, res)
  )
module.exports.cleanup = async () => cleanup()
